components/tooltip/tooltip.js

```diff
@@ -0,0 +1,238 @@
+import { clearDismissible, setDismissible } from '../../helpers/dismissible.js';
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { bodyCompactStyles } from '../typography/styles.js';
+import { getUniqueId } from '../../helpers/uniqueId.js';
+import { RtlMixin } from '../../mixins/rtl-mixin.js';
+
+class Tooltip extends RtlMixin(LitElement) {
+
+	static get properties() {
+		return {
+			delay: { type: Number },
+			disableFocusLock: { type: Boolean, attribute: 'disable-focus-lock' },
+			for: { type: String },
+			forceShow: { type: Boolean, attribute: 'force-show' },
+			showing: { type: Boolean, reflect: true },
+			state: { type: String, reflect: true }, /* Valid values are: 'info' and 'error' */
+		};
+	}
+
+	static get styles() {
+		return [bodyCompactStyles, css`
+			:host {
+				--d2l-tooltip-background-color: var(--d2l-color-ferrite); /* Deprecated, use state attribute instead */
+				background-color: var(--d2l-tooltip-background-color);
+				box-sizing: border-box;
+				color: white;
+				display: none;
+				position: absolute;
+				text-align: left;
+				white-space: normal;
+				z-index: 1000; /* position on top of floating buttons */
+			}
+
+			:host([state="error"]) {
+				--d2l-tooltip-background-color: var(--d2l-color-cinnabar);
+			}
+
+			:host([dir="rtl"]) {
+				text-align: right;
+			}
+
+			:host([showing]) {
+				display: inline-block;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+
+		this._onMouseEnter = this._onMouseEnter.bind(this);
+		this._onMouseLeave = this._onMouseLeave.bind(this);
+		this._onFocus = this._onFocus.bind(this);
+		this._onBlur = this._onBlur.bind(this);
+
+		this.delay = 0;
+		this.disableFocusLock = false;
+		this.forceShow = false;
+		this.showing = false;
+		this.state = 'info';
+
+		this._isFocusing = false;
+		this._isHovering = false;
+		this._dismissibleId = null;
+	}
+
+	get for() {
+		return this._for;
+	}
+	set for(val) {
+		const oldVal = this._for;
+		if (oldVal !== val) {
+			this._for = val;
+			this.requestUpdate('for', oldVal);
+			this._updateTarget();
+		}
+	}
+
+	get forceShow() {
+		return this._forceShow;
+	}
+	set forceShow(val) {
+		const oldVal = this._forceShow;
+		if (oldVal !== val) {
+			this._forceShow = val;
+			this.requestUpdate('force-show', oldVal);
+			this._updateShowing();
+		}
+	}
+
+	get showing() {
+		return this._showing;
+	}
+	set showing(val) {
+		const oldVal = this._showing;
+		if (oldVal !== val) {
+			this._showing = val;
+			this.requestUpdate('showing', oldVal);
+			this._showingChanged(val);
+		}
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		requestAnimationFrame(() => {
+			this._updateTarget();
+		});
+	}
+
+	disconnectedCallback() {
```

Comment:
If the tooltip is disconnected from the DOM, it looks like the listeners on the target will not be removed because they are not part of the tootip's sub-tree.  The listeners will continue to hold references to everything accessible within the handler's closure (like `d2l-tooltip`), preventing it from being cleaned up.  It is a scenario that we might not expect, but I think it could easily arise.

Reply:
Good catch, I'll add `_removeListeners`